Source/JavaScriptCore/runtime/IntlDurationFormat.cpp

```diff
@@ -470,33 +470,36 @@ static Vector<Element> collectElements(JSGlobalObject* globalObject, const IntlD
             // 3.l.i. If style is "2-digit" or "numeric", then
             case IntlDurationFormat::UnitStyle::TwoDigit:
             case IntlDurationFormat::UnitStyle::Numeric: {
-                auto formattedNumber = formatDouble(skeletonBuilder.toString());
-                RETURN_IF_EXCEPTION(scope, { });
+                // https://tc39.es/proposal-intl-duration-format/#sec-formatnumericunits
+                ASSERT(unit == TemporalUnit::Hour || unit == TemporalUnit::Minute || unit == TemporalUnit::Second);
 
-                auto formatted = formatToString(formattedNumber.get());
-                RETURN_IF_EXCEPTION(scope, { });
+                double secondsValue = duration[TemporalUnit::Second];
+                if (durationFormat->units()[static_cast<unsigned>(TemporalUnit::Millisecond)].style() == IntlDurationFormat::UnitStyle::Numeric)
+                    secondsValue = secondsValue + duration[TemporalUnit::Millisecond] / 1000.0 + duration[TemporalUnit::Microsecond] / 1000000.0 + duration[TemporalUnit::Nanosecond] / 1000000000.0;
 
-                elements.append({ ElementType::Element, unit, WTFMove(formatted), value, WTFMove(formattedNumber) });
+                bool needsFormatHours = duration[TemporalUnit::Hour] || durationFormat->units()[static_cast<unsigned>(TemporalUnit::Hour)].display() != IntlDurationFormat::Display::Auto;
+                bool needsFormatSeconds = secondsValue || durationFormat->units()[static_cast<unsigned>(TemporalUnit::Second)].display() != IntlDurationFormat::Display::Auto;
+                bool needsFormatMinutes = (needsFormatHours && needsFormatSeconds) || (duration[TemporalUnit::Minute] || durationFormat->units()[static_cast<unsigned>(TemporalUnit::Minute)].display() != IntlDurationFormat::Display::Auto);
```

Comment:
nit: I don't think you need parens around the second `||`.
```suggestion
                bool needsFormatMinutes = (needsFormatHours && needsFormatSeconds) || duration[TemporalUnit::Minute] || durationFormat->units()[static_cast<unsigned>(TemporalUnit::Minute)].display() != IntlDurationFormat::Display::Auto;
```